include/dali/core/cuda_event.h

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_CUDA_EVENT_H_
+#define DALI_CORE_CUDA_EVENT_H_
+
+#include <driver_types.h>
+#include <utility>
+#include "dali/core/unique_handle.h"
+
+namespace dali {
+
+class DLL_PUBLIC CUDAEvent : public UniqueHandle<cudaEvent_t, CUDAEvent> {
+ public:
+  DALI_INHERIT_UNIQUE_HANDLE(cudaEvent_t, CUDAEvent)
+  constexpr CUDAEvent() = default;
+
+  /// @brief Creates an event on specified device (or current device, if device_id < 0)
+  static CUDAEvent Create(int device_id = -1);
```

Comment:
Why would we need here both, factory method and a public ctor? Isn't one of them redundant?

Reply:
1. Tthe `Create` function has more logic than a simple call to `cudaEventCreate` has (the device thing).
2. There's no constructor that actually creates the event. And the create function is safer, because the handle is immediately under control of RAII. For example:
```
std::vector<CUDAEvent> events;
cudaEvent_t ev;
cudaEventCreate(&ev);
events.emplace_back(ev);  // if emplace_back throws bad_alloc, a resource (ev) is leaked
```
In contrast, this handles out-of-memory condition just fine:
```
std::vector<CUDAEvent> events;
events.push_back(CUDAEvent::Create());  // if emplace_back throws bad_alloc, a resource (ev) is leaked
```